candidategenerator.js

```diff
@@ -28,14 +28,82 @@ var wNum = 0;
 
+	options.recursiveCategory = options.recursiveCategory || "phi"; //sets the default of recursiveCategory option to "phi"
+
+	//set default root root category based on options passed
+	if(!options.rootCategory){ //root category is unspecifed
+		//recursiveCategory is specified as "w"
+		if(options.recursiveCategory === "w"){
+			options.rootCategory = "phi";
+		}
+		else{
+			options.rootCategory = "i";
+		}
+	}
+
+	//set default terminal category based on options passed
+	if(!options.terminalCategory){ //terminalCategory is unspecified
+		//terminalCategory unspecified, recursiveCategory specified as "i"
+		if(options.recursiveCategory === "i"){
+			options.terminalCategory = "phi";
+		}
+		//terminalCategory unspecified, recursiveCategory specified as "w"
+		else if(options.recursiveCategory === "w"){
+			options.terminalCategory = "Ft";
+		}
+		//neither terminalCategory nor recursiveCategory is specified, default to "w"
+		else{
+			options.terminalCategory = "w";
+		}
+	}
+
+	/* the prosodic hierarchy should include the categories specified in
+	 * options.rootCategory, options.recursiveCategory and options.terminalCategory
+	 */
+	if(options.rootCategory || options.recursiveCategory || options.terminalCategory){
+		//the specified root category should probably be highest
+		if(pCat.indexOf(options.rootCategory)<0)
+			pCat.unshift(options.rootCategory);
+		if(pCat.indexOf(options.recursiveCategory)<0 && options.rootCategory !== options.recursiveCategory){
+			//the specified recursive category should be after iota
+			try{
+				var afterI = pCat.indexOf("i") + 1;
+				pCat.splice(afterI, 0, options.recursiveCategory);
+			}
+			catch(err){
+				console.error("make sure pCat contains your recursive category.");
+			}
+		}
+		//the specified terminal category should be right above "syll"
+		if(pCat.indexOf(options.terminalCategory)<0 && options.terminalCategory !== options.recursiveCategory && options.rootCategory !== options.terminalCategory)
+			pCat.splice(-1, 0, options.terminalCategory);
+	}
+
+	if(options.rootCategory === options.recursiveCategory && options.obeysNonrecursivity){
+		alert("Warning:\nYou have instructed GEN to produce non-recursive trees and to produce trees where the root node and\nintermediate nodes are of the same category. Some of the trees GEN produces will be recursive.")
+	}
+
+	if(options.rootCategory === options.recursiveCategory && options.obeysNonrecursivity){
+		alert("Warning:\nYou have instructed GEN to produce non-recursive trees and to produce trees where the root node and\nterminal nodes are of the same category. Some of the trees GEN produces will be recursive.")
+	}
+
+	if(options.recursiveCategory === options.terminalCategory && options.obeysNonrecursivity){
+		alert("Warning:\nYou have instructed GEN to produce non-recursive trees and to produce trees where the intermediate\nnodes and the terminal nodes are of the same category. You will only get one bracketing.")
+	}
+
```

Comment:
We need a check and warning somewhere in here if recCat isn't the next category down the prosodic hierarchy from rootCat. E.g., if someone asks for trees rooted in iota with w as the recursive category, there aren't going to be any phis in the results. That is probably not what they have in mind. Every single tree will be violating Exhaustivity in a major way.

Reply:
We also need a check for situations that can create layering violations. E.g., the user specifies {rootCategory:'i', terminalCategory: 'i'}, but doesn't specify recursiveCategory. Then by default, recursiveCategory is set to 'phi' and you end up with trees where the root and leaves are iotas but the intermediate nodes are phis. This merits a warning or error.

Basically at the end of all the category setting, there needs to be a check that rootCategory >= recursiveCategory >= terminalCategory, and no levels are skipped between them.

Reply:
updated in forthcoming commit